onset_maproom/tests/test_calc.py

```diff
@@ -78,6 +202,254 @@ def test_water_balance_et_has_T():
+       [ 6.34475728],
+       [ 4.69855057],
+       [ 3.27655007],
+       [ 1.27869775],
+       [13.15848481],
+       [14.09643733],
+       [14.15885817]
+    ]
+    assert np.allclose(
+        wat_bal.soil_moisture.isel(T=slice(-20, None)),
+        expected
+    )
+
+
+def test_soil_plant_water_balance_with_et_crop():
+
+    tmin = (precip_sample() + 10).expand_dims({"Y": [14.1]})
+    tmin["Y"].attrs = dict(units="degree_north")
+    tmax = tmin * 1.4
+    temp_avg = (tmin + tmax) / 2
+    temp_amp = (tmax - tmin).clip(min=0)
+    doy = tmin["T"].dt.dayofyear
+    lat = tmin["Y"]
+    if lat.units == "degree_north":
+        lat = lat * np.pi / 180
+        lat.attrs = dict(units="radian")
+    ra = calc.solar_radiation(doy, lat)
+    p_d = xr.DataArray(
+        pd.DatetimeIndex(data=["2000-05-02", "2000-05-13"]),
+        dims=["X"],
+        coords={"X": [0, 1]},
+    ).expand_dims({"T": pd.DatetimeIndex(data=["2000-05-01"])})
+    p_d = p_d - p_d["T"]
+    kc_periods = pd.TimedeltaIndex([0, 45, 47, 45, 45], unit="D")
+    kc_params = xr.DataArray(
+        data=[0.2, 0.4, 1.2, 1.2, 0.6], dims=["kc_periods"], coords=[kc_periods]
+    )
+    wat_bal = calc.soil_plant_water_balance(
+        precip_sample(),
+        calc.hargreaves_et_ref(temp_avg, temp_amp, ra),
+        60,
+        10,
+        kc_params=kc_params,
+        p_d=p_d,
+        rho=None,
+        runoff=calc.weekly_api_runoff(precip_sample()),
+    )
+    print(wat_bal.soil_moisture)
+
+    assert (wat_bal.et_crop == wat_bal.et_crop_red).all()
+    expected = [[[55.81081862, 56.74815773]],
+       [[57.01759313, 58.28147063]],
+       [[58.09657863, 59.71744128]],
+       [[60.        , 60.        ]],
+       [[59.35689346, 59.72076937]],
+       [[58.92301271, 59.68844135]],
+       [[57.8169261 , 58.98686702]],
+       [[60.        , 60.        ]],
+       [[60.        , 60.        ]],
+       [[60.        , 60.        ]]]
+    assert np.allclose(
+        wat_bal.soil_moisture.isel(T=slice(-10, None)),
+        expected
+    )
+    
+
+def test_soil_plant_water_balance_with_et_crop_pd_none():
+
+    tmin = (precip_sample() + 10).expand_dims({"Y": [14.1]})
+    tmin["Y"].attrs = dict(units="degree_north")
+    tmax = tmin * 1.4
+    temp_avg = (tmin + tmax) / 2
+    temp_amp = (tmax - tmin).clip(min=0)
+    doy = tmin["T"].dt.dayofyear
+    lat = tmin["Y"]
+    if lat.units == "degree_north":
+        lat = lat * np.pi / 180
+        lat.attrs = dict(units="radian")
+    ra = calc.solar_radiation(doy, lat)
+    kc_periods = pd.TimedeltaIndex([0, 45, 47, 45, 45], unit="D")
+    kc_params = xr.DataArray(
+        data=[0.2, 0.4, 1.2, 1.2, 0.6], dims=["kc_periods"], coords=[kc_periods]
+    )
+    wat_bal = calc.soil_plant_water_balance(
+        precip_sample(),
+        calc.hargreaves_et_ref(temp_avg, temp_amp, ra),
+        60,
+        10,
+        kc_params=kc_params,
+        rho=None,
+        runoff=calc.weekly_api_runoff(precip_sample()),
+    )
```

Comment:
In this test (which is the only one I've looked closely at so far), `daily_rain` and `runoff` are 1D (T is the only dimension) whereas `et` is 2D (T, Y). Is this a realistic scenario? Why would some of these parameters vary by location and others not? It all works out because of a call to `broacast(water_balance)` in the main function. I'm wondering whether that `broacast` is doing something useful, or if it's just obscuring mismatched data shapes when it would be more useful to bring it to the programmer's attention.

Reply:
Yes, it is a realistic scenario as basically et can depend on crops. Calling it Y here is copy-and-past laziness leading to confusion, sorry. Typically, the climate variables will depend on space (X/Y, station, shape) and T (or S/L) (and potentially ensemble member or this kind of stuff). Then the soil dependent data (TAW) would depend on space only. Then the crop dependent data (that is kc_params and thus et_crop and beyond) could depend on "crop." A user could choose kc_params=None and input an et that is already crop dependent.

This being said, it is possible that broadcast is unnecessary. Now I am wondering if all my expands and squeezes deal with the problems that made me broadcast. Or maybe I just need to add a few. I do get a bunch of "shape mismatch" error when commenting out broadcasting.

Reply:
Did you want to rename `Y`? Resolve this conversation if you're finished with it.